pkg/cloudprovider/vsphere/route/routeprovider.go

```diff
@@ -0,0 +1,182 @@
+	var nexthops []model.RouterNexthop
+	nexthops = append(nexthops, model.RouterNexthop{IpAddress: &nodeIp})
+	staticRoute := model.StaticRoutes{
+		Network:  route.DestinationCIDR,
+		NextHops: nexthops,
+		Tags:     tags,
+	}
+	routerId := p.getRouterId(p.routerPath)
+	routeId := nameHint + "_" + route.DestinationCIDR
+	routeId = strings.ReplaceAll(routeId, "/", "_")
+	err = p.broker.tier1StaticRoutesClient.Patch(routerId, routeId, staticRoute)
+	if err != nil {
+		klog.Errorf("Creating static route %s for node %s failed: %s", routeId, nodeName, err)
+		return err
+	}
+
+	// Get realized state
+	return p.checkStaticRouteRealizedState(p.routerPath, routeId)
+}
+
+// DeleteRoute deletes Node's static route on NSXT
+func (p *RouteProvider) DeleteRoute(ctx context.Context, clusterName string, route *cloudprovider.Route) error {
+	klog.V(6).Infof("Deleting static route %s on router %s in cluster %s",
+		route.Name, p.routerPath, clusterName)
+	routerId := p.getRouterId(p.routerPath)
+	err := p.broker.tier1StaticRoutesClient.Delete(routerId, route.Name)
+	if err != nil {
+		klog.Errorf("Deleting static route %s failed: %s", route.Name, err)
+		return err
+	}
+	return nil
+}
+
+// getNodeIPAddress gets node IP address
+// The order is to choose node internal IP first, then external IP.
+func (p *RouteProvider) getNodeIPAddress(nodeName string) (string, error) {
+	node, err := p.kubeClient.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Getting node %s failed: %v", nodeName, err)
+		return "", err
+	}
+
+	addresses := make(map[corev1.NodeAddressType]string)
+	for _, addr := range node.Status.Addresses {
+		addresses[addr.Type] = addr.Address
+	}
+	var ipAddrStr string
+	if internalIP, ok := addresses[corev1.NodeInternalIP]; ok {
+		ipAddrStr = internalIP
+	} else if externalIP, ok := addresses[corev1.NodeExternalIP]; ok {
+		ipAddrStr = externalIP
+	} else {
+		return "", fmt.Errorf("Node %s has neither internal ip nor external ip", nodeName)
+	}
+	return ipAddrStr, nil
+}
+
+// checkStaticRouteRealizedState checks static route realized state
+// The check happends every 1 second and the default timeout is 10 seconds
+// Do not delete the creating static route after the timeout
+func (p *RouteProvider) checkStaticRouteRealizedState(routerPath string, routeId string) error {
+	path := routerPath + "/static-routes/" + routeId
+	limit := time.Now().Add(config.RealizedStateTimeout)
+	for time.Now().Before(limit) {
+		time.Sleep(config.RealizedStateSleepTime)
+		list, err := p.broker.realizedEntitiesClient.List(path)
+		if err != nil {
+			return fmt.Errorf("Get route %s realized state failed: %s", path, err)
+		}
+		for _, resource := range list.Results {
+			if resource.IntentPaths[0] == path && resource.State == config.RealizedState {
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("Waiting for static route %s is timeout", path)
+}
+
+// getRouterId returns router ID from router path
```

Comment:
If we only care about the router ID can we just have the config require router ID instead of the path?

Reply:
I'm thinking if we support Tier0 router later, we could use path to identify it is Tier1 or Tier0 router

Reply:
noted, thanks